src/v/v8_engine/instance.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sstring.hh>
+#include <seastar/core/temporary_buffer.hh>
+
+#include <v8.h>
+
+namespace v8_engine {
+
+class instance {
+public:
+    explicit instance(size_t);
+
+    instance(const instance& other) = delete;
+    instance& operator=(const instance& other) = delete;
+    instance(instance&& other) = default;
+    instance& operator=(instance&& other) = default;
+
+    ~instance();
+
+    ss::future<> init(ss::sstring, ss::sstring);
+    ss::future<> run(ss::temporary_buffer<char>& data);
+
+    class exception final : public std::exception {
```

Comment:
Why exception class is nested? The caller will have to write something like 

```
} catch (v8_engine::instance::exception& e) { 
```
to catch the exception, also the name 'exception' is too generic

Also, I'm wondering, does v8 throw its own exceptions? Or it only uses error codes?

Reply:
It uses specific class for exception in v8 `v8::TryCatch`. It does not throw it to c++ code. I used check for exception status inside v8 and throwed it to `ss::future`

Reply:
Yes, I want to write `v8_engine::instance::exception`. Is it better is I rename it tot instance_exception and and make it not nested?

Reply:
the name is really is not a big deal
v8::TryCatch is likely used somewhere else, isn't it? If this is the case the code can benefit if this mechanism will be reusable, for instance, you can create a function that gets v8::TryCatch and turns it into C++ exception. Nested exception can only be used inside one class.

Reply:
Done